File: src/main/java/mt/edu/um/cf2/jgribx/GribRecordIS.java
/*
 * ============================================================================
 * JGribX
 * ============================================================================
 * Written by Andrew Spiteri <andrew.spiteri@um.edu.mt>
 * Adapted from JGRIB: http://jgrib.sourceforge.net/
 * 
 * Licensed under MIT: https://github.com/spidru/JGribX/blob/master/LICENSE
 * ============================================================================
 */
package mt.edu.um.cf2.jgribx;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.util.Arrays;
import mt.edu.um.cf2.jgribx.grib2.ProductDiscipline;


/**
 * A class that represents the indicator section (IS) of a GRIB record.
 *
 * 01/01/2001   Benjamin Stark          first version
 * 16/09/2002   Richard D. Gonzalez     modified to indicate support of GRIB edition 1 only
 * 24/05/2017   Andrew Spiteri          added support for GRIB2
 *
 */

public class GribRecordIS
{

   /**
    * Length in bytes of GRIB record.
    */
   protected long recordLength;

   /**
    * Length in bytes of IS section.
    * Section recordLength differs between GRIB editions 1 and 2
 Currently only GRIB edition 1 supported - recordLength is 8 octets/bytes.
    */
   protected int length;

   /**
    * Edition of GRIB specification used.
    */
   private int edition;

   private static String PATTERN = "GRIB";

   protected ProductDiscipline discipline;

   private long byteOffset;

    /**
     * Constructs a {@link GribRecordIS} object from a buffered input stream.
     *
     * @param in {@link BufferedInputStream} containing IS content
     * @return 
     * @throws NotSupportedException 
     * @throws mt.edu.um.cf2.jgribx.NoValidGribException 
     * @throws IOException           if stream can not be opened etc.
     */
    public static GribRecordIS readFromStream(GribInputStream in) throws NotSupportedException, NoValidGribException, IOException
    {
        GribRecordIS is = new GribRecordIS();
        is.byteOffset = in.getAbsoluteBytePosition();
        byte[] octets = new byte[16];
        in.read(octets, 0, 8);
        
        String startCode = new String(Arrays.copyOfRange(octets, 0, 4));
        if (!startCode.equals(PATTERN))
        {
            Logger.println("Start code not recognised: " + startCode, Logger.INFO);
            throw new NoValidGribException("Record does not have a valid GRIB header");
        }
      
        // check GRIB edition number
        is.edition = octets[7];
        switch (is.edition)
        {
            case 1:
                is.length = 8;
                break;
            case 2:
                is.discipline = new ProductDiscipline(octets[6]);
                is.length = 16;
                in.read(octets, 8, 8);
                break;
            default:
                throw new NotSupportedException("GRIB edition " + is.edition +
                " is not yet supported");
        }
      
        // recordLength of GRIB record
        if (is.edition == 1)
            is.recordLength = Bytes2Number.bytesToLong(Arrays.copyOfRange(octets, 4, 7));
        else if (is.edition == 2)
            is.recordLength = Bytes2Number.bytesToLong(Arrays.copyOfRange(octets, 8, 16));
        
        return is;
   }

    /**
     * Get the byte offset of the start of the section relative to the start of the file
     * @return The byte offset
     */
   public long getByteOffset() { return byteOffset; }

    public ProductDiscipline getDiscipline()
    {
        return discipline;
    }

   /**
    * Get the byte recordLength of this GRIB record.
    *
    * @return recordLength in bytes of GRIB record
    */
   public long getRecordLength()
   {
      return this.recordLength;
   }

   /**
    * Get the byte recordLength of the IS section.
    *
    * @return recordLength in bytes of IS section
    */
   public int getLength()
   {
      return this.length;
   }

    /**
     * Get the edition of the GRIB specification used.
     *
     * @return edition number of GRIB specification
     */
    public int getGribEdition()
    {
       return this.edition;
    }

    /**
     * Seeks the location of the next IS header.
     * If a valid header is found, the input stream is repositioned to the
     * position just before the found header.
     * @param in GRIB input stream to read from 
     * @throws IOException 
     */
    public static void seekNext(GribInputStream in) throws IOException
    {
        int nBytesConsumed = in.seekBytePattern(PATTERN.getBytes(), false);
        int nBytesSkipped = nBytesConsumed - PATTERN.length();

        if (nBytesSkipped > 0)
        {
            Logger.println("Extra " + nBytesSkipped + " bytes were found between end of last record "
                    + "and start of next record", Logger.WARNING);
        }
    }

   /**
    * Get a string representation of this IS.
    *
    * @return string representation of this IS
    */
   @Override
   public String toString()
   {

      return "    IS section:" + '\n' +
            "        Grib Edition " + this.edition + '\n' +
            "        length: " + this.recordLength + " bytes";
   }

}

File: src/main/java/mt/edu/um/cf2/jgribx/GribRecordIS.kt
/*
 * ============================================================================
 * JGribX
 * ============================================================================
 * Written by Andrew Spiteri <andrew.spiteri@um.edu.mt>
 * Adapted from JGRIB: http://jgrib.sourceforge.net/
 * 
 * Licensed under MIT: https://github.com/spidru/JGribX/blob/master/LICENSE
 * ============================================================================
 */
package mt.edu.um.cf2.jgribx

import mt.edu.um.cf2.jgribx.api.GribSection
import mt.edu.um.cf2.jgribx.grib2.ProductDiscipline
import java.io.IOException

/**
 * ### Section 0: Indicator Section
 *
 * #### GRIB 1
 *    | Octet | # | Value                                                                   |
 *    |-------|---|-------------------------------------------------------------------------|
 *    | 1-4   | 4 | "GRIB" (coded according to the International Alphabet No. 5)            |
 *    | 5-7   | 3 | Total length of GRIB message in octets (including Section 0)            |
 *    | 8     | 1 | GRIB Edition Number (here 1)                                            |
 *
 * #### GRIB 2
 *    | Octet | # | Value                                                                   |
 *    |-------|---|-------------------------------------------------------------------------|
 *    | 1-4   | 4 | "GRIB" (coded according to the International Alphabet No. 5)            |
 *    | 5-6   | 2 | Reserved                                                                |
 *    | 7     | 1 | Discipline – GRIB Master Table Number (see Code Table 0.0)              |
 *    | 8     | 1 | GRIB Edition Number (here 2)                                            |
 *    | 9-16  | 8 | Total length of GRIB message in octets (including Section 0)            |
 *
 * A class that represents the indicator section (IS) of a GRIB record.
 *
 * 01/01/2001   Benjamin Stark          first version
 * 16/09/2002   Richard D. Gonzalez     modified to indicate support of GRIB edition 1 only
 * 24/05/2017   Andrew Spiteri          added support for GRIB2
 *
 * @param gribEdition Edition of GRIB specification used.
 * @param discipline Discipline – GRIB Master Table Number (see [Code Table 0.0](https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_table0-0.shtml))
 * @param messageLength Length in bytes of GRIB record.
 *
 * @author Benjamin Stark
 * @author Richard D. Gonzalez
 * @author Andrew Spiteri
 * @author Jan Kubovy [jan@kubovy.eu]
 */
class GribRecordIS internal constructor(var gribEdition: Int,
										var discipline: ProductDiscipline?,
										internal var messageLength: Long) : GribSection {

	companion object {
		/**
		 * Constructs a [GribRecordIS] object from a buffered input stream.
		 *
		 * @param gribInputStream [GribInputStream] containing IS content
		 * @return
		 * @throws NotSupportedException
		 * @throws mt.edu.um.cf2.jgribx.NoValidGribException
		 * @throws IOException           if stream can not be opened etc.
		 */
		fun readFromStream(gribInputStream: GribInputStream): GribRecordIS {
			val startCode = gribInputStream.readString(4) // [1-4]
			if (startCode != "GRIB") {
				Logger.info("Start code not recognised: ${startCode}")
				throw NoValidGribException("Record does not have a valid GRIB header")
			}

			// check GRIB edition number
			val octets = gribInputStream.peek(4)
			val gribEdition = octets[3].toInt()
			val discipline: ProductDiscipline?
			val recordLength: Long
			when (gribEdition) {
				1 -> {
					discipline = null
					recordLength = gribInputStream.readLong(3) // [5-7]
					gribInputStream.skip(1) // [8]
				}
				2 -> {
					gribInputStream.skip(2) // [5-6]
					discipline = gribInputStream.readUInt(1) // [7]
							.let { ProductDiscipline.VALUES[it] ?: TODO("Discipline ${it} not implemented") }
					gribInputStream.skip(1) // [8]
					recordLength = gribInputStream.readLong(8) // [9-16]
				}
				else -> throw NotSupportedException("GRIB edition ${gribEdition} is not yet supported")
			}

			return GribRecordIS(gribEdition, discipline, recordLength)
		}

		/**
		 * Seeks the location of the next IS header.
		 * If a valid header is found, the input stream is repositioned to the position just before the found header.
		 *
		 * @param gribInputStream GRIB input stream to read from
		 * @return Whether an IS header was found till end of stream or not
		 */
		fun seekNext(gribInputStream: GribInputStream): Boolean {
			var bytesSkipped = 0
			var found = false
			while (gribInputStream.available() > 0) {
				found = gribInputStream.peek(4).contentEquals("GRIB".toByteArray())
				if (found) break
				gribInputStream.skip(1) // skip 1 byte
				bytesSkipped++
			}
			if (bytesSkipped > 0) {
				Logger.warning("Extra ${bytesSkipped} bytes were found between end of last record and start of next record")
			}
			return found
		}
	}

	/**
	 * Length in bytes of IS section. Section recordLength differs between GRIB editions 1 and 2. Currently only GRIB
	 * edition 1 supported - recordLength is 8 octets/bytes.
	 */
	override val length: Int
		get() = when (gribEdition) {
			1 -> 8
			2 -> 16
			else -> throw NotSupportedException("GRIB edition ${gribEdition} is not yet supported")
		}


	override fun writeTo(outputStream: GribOutputStream) {
		Logger.debug("Writing GRIB${gribEdition} Indicator Section (IS) - ${length} bytes (total: ${messageLength} bytes)")
		outputStream.write("GRIB".toByteArray())
		when (gribEdition) {
			1 -> {
				outputStream.writeLong(messageLength, bytes = 3) // [5-7] Record length
				outputStream.writeUInt(gribEdition, bytes = 1) // [8] GRIB Edition
			}
			2 -> {
				outputStream.write(byteArrayOf(0, 0)) // [5-6] Reserved
				outputStream.writeUInt(discipline?.value ?: ProductDiscipline.MISSING, bytes = 1) // [7] Discipline
				outputStream.writeUInt(gribEdition, bytes = 1) // [8] GRIB Edition
				outputStream.writeLong(messageLength, bytes = 8) // [9-16] Record length
			}
			else -> throw Exception("Unsupported GRIB edition ${gribEdition}")
		}
	}

	override fun equals(other: Any?) = this === other
			|| other is GribRecordIS
			&& length == other.length
			&& gribEdition == other.gribEdition
			&& discipline == other.discipline
			&& messageLength == other.messageLength

	override fun hashCode(): Int {
		var result = length
		result = 31 * result + gribEdition
		result = 31 * result + (discipline?.hashCode() ?: 0)
		result = 31 * result + messageLength.hashCode()
		return result
	}

	override fun toString(): String = listOfNotNull(
			"GRIB IS section:",
			"\tGrib edition ${gribEdition}",
			discipline?.let { "\tDiscipline: ${it}" },
			"\tLength: ${length}",
			"\tRecord length: ${messageLength} bytes")
			.joinToString("\n")
}
